### PR TITLE
feat(bitbucket): add support for pagelen

### PR DESCRIPTION
**Pull request by setchy**
## Changes

- Add support for optional paglen
- Add max pagelen if non provided and requesting paginated data
- Update pagination logic to use next url as-is previous response

## Context

Pre-req for #22272

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
missing code coverage

**Comment by setchy** [comment]
> missing code coverage

🙈 - oops.  all good now :)

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/util/http/bitbucket.ts [hunk 1]
```diff
@@ -60,25 +63,26 @@ export class BitbucketHttp extends Http<BitbucketHttpOptions> {
...
+    return false;
   }
 
-  return resolvedURL.searchParams.get('page');
+  return !is.nullOrUndefined(resolvedURL.searchParams.get('pagelen'));
 }
 
-function getNextPagePath(path: string, nextPage: string): string | null {
+function addPagelenToPath(path: string, pagelen: number): string {
   const resolvedURL = parseUrl(resolveBaseUrl(baseUrl, path));
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/util/http/bitbucket.ts]: lmk if the updated code is what you had in mind.  handling the unlikely null case feels like a smell to me. interested in alternate suggestions

File: lib/util/http/bitbucket.ts [hunk 2]
```diff
@@ -60,27 +70,13 @@ export class BitbucketHttp extends Http<BitbucketHttpOptions> {
...
-  // istanbul ignore if
-  if (is.nullOrUndefined(resolvedURL)) {
-    return null;
-  }
-
-  resolvedURL.searchParams.set('page', nextPage);
-
-  return resolvedURL.toString();
+function addPagelenToPath(url: URL, pagelen: number): URL {
+  url.searchParams.set('pagelen', pagelen.toString());
```
Comment (viceice) [inline lib/util/http/bitbucket.ts:78]: i think this should now inlined, it's only used once

Comment (setchy) [reply lib/util/http/bitbucket.ts]: Agreed.  inlined

**Review by rarkins — DISMISSED** [review]

File: lib/util/http/bitbucket.ts [hunk 3]
```diff
@@ -25,29 +29,35 @@ export class BitbucketHttp extends Http<BitbucketHttpOptions> {
...
-        // istanbul ignore if
-        if (is.nullOrUndefined(nextPath)) {
-          break;
-        }
+    if (opts.paginate && isPagedResult(result.body)) {
+      const resultBody = result.body as PagedResult<T>;
 
+      let nextURL = resultBody.next;
+
+      while (is.nonEmptyString(nextURL)) {
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/util/http/bitbucket.ts:52]: To confirm,  up to 100 pages each with a max pagelen of 100?

So a max of 10,000 results?

[the remark this answers is not in the document]
Comment (setchy) [reply lib/util/http/bitbucket.ts]: Added

**Review by viceice — APPROVED** [review]

**Comment by setchy** [comment]
@rarkins - apologies, your approval got clearer.  🙏

**Review by secustor — APPROVED** [review]

**Pull request merged**